src/collapse/xctrace.rs

```
@@ -0,0 +1,439 @@
+            other => matches!(self, Self::Other(tag) if tag.as_bytes() == other),
+        }
+    }
+}
+
+#[derive(Default)]
+struct TraceQueryResultState {
+    nodes: Vec<Node>,
+}
+
+#[derive(Default)]
+struct NodeState {
+    rows: Vec<Row>,
+}
+
+#[derive(Default)]
+struct RowState {
+    backtrace: Option<Arc<Backtrace>>,
+}
+
+#[derive(Default)]
+struct BacktraceState {
+    id: u64,
+    frames: Vec<Arc<Frame>>,
+}
+
+#[derive(Default)]
+struct FrameState {
+    id: u64,
+    name: Vec<u8>,
+}
+
+struct Node {
+    rows: Vec<Row>,
+}
+
+struct Row {
+    backtrace: Arc<Backtrace>,
+}
+
+struct Backtrace {
+    id: u64,
+    frames: Vec<Arc<Frame>>,
+}
+
+struct Frame {
+    id: u64,
+    name: Vec<u8>,
+}
+
+impl Backtrace {
+    fn to_folded(&self) -> Vec<u8> {
+        let mut folded = Vec::new();
+        let mut first = Some(());
+        // Because stack frames are arranged from top to bottom in xctrace's
+        // output, here we use `.rev(`.
+        for frame in self.frames.iter().rev() {
+            if first.take().is_none() {
+                folded.push(b';');
+            }
+            let frame_name = String::from_utf8_lossy(&frame.name);
+            let frame_name = fix_partially_demangled_rust_symbol(&frame_name);
+            folded.extend(frame_name.as_bytes().iter().copied());
+        }
+        folded
+    }
+}
+
+/// Unescapes the text in xml exported from xctrace.
+fn unescape_xctrace_text(text: Cow<'_, [u8]>) -> Vec<u8> {
+    // xctrace shouldn't give us invalid xml text here, therefore
+    // we don't expect the error branch being hit:
+    //
+    // `quick_xml::escape::unescape` will error out if the input is not a valid xml text:
+    // https://github.com/tafia/quick-xml/blob/0793d6a8d006cb5dabf66bf2a25ddbf198305b46/src/escape.rs#L253
+    quick_xml::escape::unescape(&String::from_utf8_lossy(&text))
+        .expect("Invalid xml text from xctrace, which is not expected")
```

Ah, we probably don't want to panic here (which is what `expect` will do). Instead, I think we should make this function return a `Result` and propagate an `Err` here.

---

Do we really want to propagate the error if we have no plan of handling it anyway?

---

Yes — we'd rather surface it to the caller who can then choose what to do (including panic), rather than _force_ a program abort.

---

This situation should only occur if there is a bug in xctrace, which is rare, so I choose to panic here.

From inferno's perspective, it might not want to panic over abnormal input(which is not a bug in inferno), I am okay with it.